### ISSUE TITLE
Kodi 18.0 stutters when playing DVD files (VIDEO_TS) on Fire TV

### ISSUE THREAD

**Issue by robertdahlem** [xbmc/xbmc#15405]
### Describe the bug
I am running Kodi 18.0 as a fresh install on an Amazon Fire TV 3rd gen (2017).
When I try to play DVD files (VIDEO_TS.IFO) the movie stutters.
It doesn't matter if I install over 17.6 or uninstall 17.6 beforehand for a fresh install.
It doesn't matter if I play from SMB (Gigabit ethernet, so no speed problem) or from a local directory (/sdcard/Movies).
This worked fine with Kodi 17.6 before.
Other movie types (i.e. MKV) are running fine.

## Expected Behavior
DVD videos play smoothly.

## Actual Behavior
DVD videos start to play, then stutter. Audio stutters a little later (after 10 seconds).

### To Reproduce

1.  Use Amazon Fire TV 3rd gen (2017)
2. Copy a DVD structure below /sdcard/movies:
AUDIO_TS
VIDEO_TS
VIDEO_TS\VIDEO_TS.IFO
VIDEO_TS\VTS_01_0.IFO
VIDEO_TS\VTS_01_1.VOB
VIDEO_TS\VTS_01_2.VOB
...
3. Install Kodi 18.0. Add /sdcard/movies as Data source
4. Play the movie. It stutters.

### Debuglog
The debuglog can be found here: https://paste.kodi.tv/utojotibow.kodi

### Your Environment
Used Operating system:

 - [x] Android

 - Operating system version/name: Amazon AFTN with Android 7.1.2 API level 25
 - Amazon software version: Fire OS 6.2.5.8 (NS6258/1607)
 - Kodi version: 18.0 Git:20190128-d81c34c

**Comment by robertdahlem** [comment]
> On your FireTV try to disable "Mediacodec and Mediacodec Surface" before playing the DVD please.

This indeed helped. Thank you.

> mediacodec is as stable as the vendor implementation :-) 

I am sorry but I fail to make sense of this.

 - If "Mediacodec" is something within Kodi then Kodi 18 introduced a bug that wasn't present in 17.6. In this case a reference to the vendor implementation would not apply.
 - If "Mediacodec" is something within Fire TV then it is still the same as it was with Kodi 17.6. Why would it be less stable when used with Kodi 18.0?

**Comment by fritsch** [comment]
No - in v17. Mediacodec was disabled by default for DVDs. Some forum users forced us to enable it by default - screaming loud and shout, you know.

**Comment by robertdahlem** [comment]
Well, now you have me screaming and shouting ...

How about a switch like "Disable Mediacodec when playing DVDs"?

**Comment by fritsch** [comment]
Nope - we rather want that FireTV people just fix their decoder :-) ... and this won't happen if we add workaround after workaround.

**Comment by robertdahlem** [comment]
> We can expect a lot of thing from Amazon. We also expect users to understand why we do certain things (do not workaround broken firmware by adding stupid switches).

Actually, we are at a point where a perfectly working feature has been removed from Kodi between 17.6 and 18.0. I prefer to not throw around the word "stupid" but deleting something that made Kodi work out of the box for anything thrown at it and instead have users fiddle around in advanced settings does not look like a completely mature and understandable decision to me.

> We did not make the decision for you to buy an Amazon device that has clearly not been designed to play such files. So i suggest to find some other device then.

I do not know how I deserve to be treated in that manner. I asked politely about the circumstances and after I read the facts that have been told here I asked politely to add a switch that brings back a deleted feature. Also I disagree about the FireTV "clearly not been designed to play such files". Kodi on FireTV played them perfectly until 17.6. It does play them perfectly in 18.0 if I fiddle around in advanced settings. My request is to reduce the necessary fiddling.

The Wiki (https://kodi.wiki/view/Supported_hardware) says: "**The main aim for the Android port of Kodi is to foremost target media-players/set-top-boxes/sticks that connect to a large screen television** and uses a standard remote control as its main interface device." That sounds like the mere definition of "FireTV" to me. We all know that Android is a heterogeneous market. Nearly every device has this or that shortcoming. No reason to throw the FireTV out of the window.

**Comment by fritsch** [comment]
Your argumentation is fully correct. Perhaps amazon did not even try to fix it, cause we had it off in the past ... pure speculations. I don't like when users suffer though.

Okay with a custom v18 build that disable Mediacodec for mpeg-2?

**Comment by fritsch** [comment]
http://mirrors.kodi.tv/test-builds/android/arm/kodi-20190203-6958f4c5-androidswmpeg2-armeabi-v7a.apk

**Comment by Markman-B** [comment]
I'm probably facing the same issues as Robert Dahlem on my Beelink GT1 with stuttering dvds. I'll try the test build and will report the results later.

**Comment by MisterT87** [comment]
I also have the same problem but on .avi files (mpeg4 - xvid codec). Disabling mediacodec works fine (stuttering is gone) - But more cpu load and way more battery usage coz of disabling hardware acceleration.

Also tested that Kodi build from @fritsch - but didnt solve my problem (sure, its another codec).
But generally with the build from @fritsch other codecs are playing better (I think) - less cpu load on activated hardware acceleration

**Comment by fritsch** [comment]
I can disable for mpeg4 < 720 width again. Will upload a tesbuild tonight.

Am Mo., 4. Feb. 2019, 13:25 hat Markman-B <notifications@github.com>
geschrieben:

**Comment by MisterT87** [comment]
Nice thank you (just for interest: the only mpeg4 files I have tested was SD quality).
So like mpeg2 before - you only disable hardware acceleration for this two codecs ?
Is there anyway to fix later ? Sure the problem is gone for now but the downside is more battery usage (cpu load).
So i think for future proof is to fix it not only disabling it or is there some misunderstood of me ?

**Comment by fritsch** [comment]
No problem. Let's try.

Am Mo., 4. Feb. 2019, 16:06 hat MisterT87 <notifications@github.com>
geschrieben:

**Comment by MisterT87** [comment]
sure - can you link it here if build is ready ? :)

**Comment by fritsch** [comment]
Will result in very same folder in roughly 60 minutes. More I cannot do for you. Bye :-)

**Comment by wrxtasy** [comment]
Finally some sanity after @robertdahlem's well constructed argument.

@fritsch is this just for mpeg2 DVD's or all mpeg2 content <720 ?

Any chance we can get All < 800 mpeg2 content SW decoded by adding:
https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp#L435
```
    if (hints.width <= 800)       
      goto FAIL;
```
If you include your powrevert, that also then works well on ARM platforms last time I tested:
https://github.com/fritsch/xbmc/compare/e37c42572b2137b3336a0d242227409430129acb...powrevert

Basically I'm trying to get 25i/720x576 mpeg2 DVD's and broadcast TV SW decoded and deinterlaced on Android ARM.

Current test .apk's above do not default to SW decoding for 25i/720x576 mpeg2 TV.

**Comment by fritsch** [comment]
Better use this: https://github.com/fritsch/xbmc/commits/androidswmpeg2 - the rest is only hackery ...

**Comment by wrxtasy** [comment]
Any chance of a test version. For some unknown reason I only get half motion h264 deinterlacing when I personally custom compile, plus run out of compiler resources here locally.

**Comment by wrxtasy** [comment]
@fritsch can you try this instead for settings options:
https://paste.kodi.tv/tujoziraha.kodi

about to runtime test on AML Linux Leia...

**Comment by fritsch** [comment]
No time. Feel free to fix it up or send a PR. Remember: Android specific
settings should not bloat settings.xml which is why I tried to move it to
android.xml where it's place is

Whenever you send a PR to my repo I will start a build.



Am Di., 5. Feb. 2019, 10:36 hat wrxtasy <notifications@github.com>
geschrieben:

**Comment by wrxtasy** [comment]
Ok, cool. PR inbound to test shortly.

**Comment by CiNcH83** [comment]
> Some forum users forced us to enable it by default - screaming loud and shout, you know.

For a good reason. SW rendering performance has dramatically been degraded. With that approach, neither HW nor SW would have worked flawlessly on the Fire TV.


> Actually, we are at a point where a perfectly working feature has been removed from Kodi between 17.6 and 18.0.

HW decoding DVD/MPEG-2 still works on many devices. And for those devices where it does not work, MediaCodec can be disabled manually. So it still works. Forcing SW decoding means far (and I really mean far) inferior scaling and deinerlacing on modern TVs. And there has been no way of enabling HW decoding in 17 Krypton. So what's the better solution?

With this android.xml approach, those switches are not accessible by the user either as far as I can see.

What about @peak3d's mentioned [blacklist approach](https://github.com/xbmc/xbmc/pull/15286#issuecomment-456451064)?

**Comment by robertdahlem** [comment]
> > Some forum users forced us to enable it by default - screaming loud and shout, you know.
> 
> For good reason. SW rendering performance has dramatically been degraded. With that approach, neither HW nor SW would have worked flawlessly on the Fire TV.

That is not my observation. 17.6 silently switched off Mediacodec for DVDs and fell back to SW rendering with flawless performance. 18.0 did not automatically switch off Mediacodec, DVDs stutter. When I manually switched off Mediacodec DVDs worked. Fritsch's version (18.1 RC1) returned the 17.6 behaviour and DVDs worked out of the box.

To sum that up: on todays FireTV DVDs do not work with HW rendering and work with SW rendering.

> HW decoding DVD/MPEG-2 still works on many devices.

I do not dispute that.

> And for those devices where it does not work, MediaCodec can be disabled manually. So it still works.

That leads to a user annoyance: we have at least on device (the FireTV) where you would need to manually switch on or off Mediacodec depending on the type of content you watch.

> Forcing SW decoding means far (and I really mean far) inferior scaling and deinerlacing on modern TVs.

I do not advocate "forcing SW decoding". What I request is a configuration option that allows me to use Mediacodec in the default case and to switch it off explicitly for DVD material. The former we have already, the latter is my request. Ideally it would switch off DVD HW rendering automatically (like in 17.6) but only if FireTV hardware is detected. One can dream ...

I do understand that additional switches are a disputed topic. From my point of view switches come into play when you can't make a sensible choice fitting all cases. Forcing the user to change a setting each time they change content type is not a sensible choice. Deliberately sabotaging DVD playback on the FireTV to force Amazon to fix their Mediacodec will produce nothing: you can force Amazon to nil. Instead, users will only see that DVD playback does not work with Kodi 18.0 on FireTV and will stay with 17.6.

> What about @peak3d's mentioned [blacklist approach](https://github.com/xbmc/xbmc/pull/15286#issuecomment-456451064)?

That is one way to go: blacklist Mediacodec for combinations of hardware and content type that are known to malfunction. I would still opt to make that user configurable so people can experiment (i.e. when a new FireOS comes out).

**Comment by CiNcH83** [comment]
> When I manually switched off Mediacodec DVDs worked.

True. Final 18 fixed the SW rendering bottleneck, or at least mitigated it, enabling proper playback of 25fps (incl. interlaced with deinterlace-half) material. All I wanted to say was that people have been asking for HW decoding when SW was badly broken, countering @fritsch's alternative facts.

I think that the real issue is understood now and that it might be safe to re-enable SW for DVD and MPEG-4 ASP SD by default. Probably all of today's Android TV devices can cope with that. SW decoding has quite some downsides though (performance- and quality-wise) and since many devices properly support HW decoding, forcing SW would't be a good solution IMHO (which I understand you are not advocating).

**Comment by robertdahlem** [comment]
> When I manually switched off Mediacodec DVDs worked.

> True. Final V18 fixed the SW rendering bottleneck, or at least mitigated it, enabling proper playback of 25fps (incl. interlaced with deinterlace-half) material.

I just downgraded to 17.6 to check that. DVDs play fine if Mediacodec is switched on or off. So, no: V18 fixed nothing with SW rendering that needed to be fixed, at least not on FireTV.

> All I wanted to say was that people have been asking for HW decoding when SW was badly broken, countering @fritsch's alternative facts.

"Alternative facts"? Do you consider this a contribution to the discussion which should be taken seriously? Because I have my doubts.

**Comment by CiNcH83** [comment]
> So, no: V18 fixed nothing with SW rendering that needed to be fixed

I meant V18 final relative to earlier V18 RC/nightly builds.

**Comment by wrxtasy** [comment]
> Anyway, doesn't matter indeed. A good solution needs to be found.

Here you go.
There is a solution but it's unacceptable to the Kodi codebase gatekeepers. I cannot be bothered trying to implement an alternative solution at this time, nor spend time Android compiling.

It uses the same user configurable HW SW switching as SPMC 16.x / 17:
https://github.com/fritsch/xbmc/pull/13

You will also need this set of @fritsch patches for better ARM rendering performance:
https://github.com/fritsch/xbmc/commits/androidswmpeg2

So compile away and Android AML and Fire TV users will have a workable solution.

**Comment by koying** [comment]
IIRC, the issue is very specific to MPEG2 from dvd, which has some peculiarity (that I can't remember right now) that some h/w decoders chokes upon.

Just bring back "hints.dvd" from https://github.com/xbmc/xbmc/blob/Krypton/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp#L374 , one way or another ;)

**Comment by davilla** [comment]
lol

**Comment by CiNcH83** [comment]
People screamed because SW wasn‘t feasible anymore due to a bottleneck in the GLES rendering path.

SW means far inferior PQ. HW deinterlacers and scalers are so much better.

**Comment by koying** [comment]
Oh, so Kodi went to the quick hack rather than fixing the GLES rendering path? Hehe...
I think I remember the issue is specific to DVD on filesystem, btw. Pretty sure ISO's didn't have the issue.

**Comment by peak3d** [comment]
A good example is the bugs bunny NTSC iso (VOB container + mpeg2)
Does not work at all with h/w on both AFTV devices, but they stutter differently.

This ISO plays well on WETEK Hub and shield b.t.w.

**Comment by alekseyzimin** [comment]
Hi, thank you fritsch for your build!  However it has a problem -- the hardware acceleration options do not show under "Player" menu.  I modified your build and re-compiled Kodi, putting the settings back into settings.xml instead of android.xml.  This fixed the problem and now there are user-configurable switches to turn software decoding on/off/HD only for mpeg2, mpeg4 and h.264. Here is my build apk:
ftp://ftp.ccb.jhu.edu/pub/alekseyz/kodiapp-armeabi-v7a-debug.apk

I tested this apk on my fire TV 3 and everything works perfectly, DVD, HD, 4K, HDR, autoframerate, etc.

This is my first time buiding Kodi.  I am concerned about apk size -- it is over90Mb, whereas your apk is about 60Mb.  I suspect the standard "make apk" command maked apk with debugging information -- how can I make a smaller ~60Mb apk?

**Comment by fritsch** [comment]
Yes. I know, just one number is wrong in android.xml. But I stopped caring.

There are guys in this thread knowing better anyways, let's wait on them.

Talk is cheap, you know :-)

Am Mi., 20. Feb. 2019, 03:57 hat Aleksey Zimin <notifications@github.com>
geschrieben:

**Comment by alekseyzimin** [comment]
I would just like to have a working Kodi that plays everything on my Fire TV :-)  Now I have it and everything works to my satisfaction! 
Out of curiosity, which number is wrong in your android.xml?

**Comment by peak3d** [comment]
https://github.com/fritsch/xbmc/commit/188049a8d43a9d2eee49ad5221ae1ea9fbf68c3f#diff-b079500169692009cb0ec16672c8d29aR34

must be group id="3"

**Comment by alekseyzimin** [comment]
I have cloned the Kodi master branch, and introduced the same changes into settings.xml, Settings.cpp, Settings.h, and DVDVideoCodecAndroidMediaCodec.cpp.  Compiled and tested -- works just fine.  Plays all DVD ISOs, Blu-ray ISOs, autoframerate, 4K, HDR, all works.  I am running GUI at 720p to allow TV to do the upscaling in most cases, because Kodi does not seem to be able to lower its resolution, i.e. when GUI is set to 1080p, it will upscale all videos to 1080p. Kodi still upscales 480p and 576p files, but this looks fine.  
Here is the apk:

ftp://ftp.ccb.jhu.edu/pub/alekseyz/kodi/kodi18.1master-armeabi-v7a-androidswmpeg2-mod.apk

the apk compiled from fritsch's fork branch androidswmpeg2 with my modifications is here:

ftp://ftp.ccb.jhu.edu/pub/alekseyz/kodi/kodi18.1RC1-armeabi-v7a-fritsch-androidswmpeg2-mod.apk

As far as I can see these apks werk the same on my Fire TV 3 pendant.  I am posting this apk for those users who are unable to compile Kodi, but would like a working alternative 18.1 release (or 18.2RC) with SW decoder switches.  It still puzzles me why Kodi developers do not want to include these switches into the main build -- they do not hurt anything, and they help many users of popular Fire TV and other non-Kodi-oriented hardware.

**Comment by jerethi** [comment]
@alekseyzimin 

Thanks for the new builds. Quick question - I wasn't following the difference between the two apks. What is the difference?

**Comment by MartijnKaijser** [comment]
Stop posting these builds and whatever. They bring absolutely nothing to this created issue ticket. If this continues i will close and lock it

**Comment by alekseyzimin** [comment]
With all due respect, there would be no need for these builds and for this whole thread if developers of team Kodi agreed to allow users to choose to use SW decoding for codecs (mostly mpeg2 and mpeg4, and interlaced h264) that are poorly supported on popular, but NOT designed for Kodi use, platforms like Amazon Fire TV.  I do not think that Amazon cares to support mpeg2 and interlaced formats properly because they have no use for it, and so one would not expect them to work on proper implementation.

**Comment by MartijnKaijser** [comment]
Go buy a different device is my advice

**Comment by Syxton** [comment]
I agree with @alekseyzimin.  Amazon has no incentive to fix this, but team KODI does.  KODI should be user driven, and I don't understand why you wouldn't accept code that is already available in order to make all users happy with your platform?  Who would be upset by the ability to make all devices just work?

**Comment by robertdahlem** [comment]
> Go buy a different device is my advice

@MartijnKaijser, do you still need to be reminded of https://kodi.wiki/view/Supported_hardware ?

"**The main aim for the Android port of Kodi is to foremost target media-players/set-top-boxes/sticks that connect to a large screen television** and uses a standard remote control as its main interface device."" In short: "Amazon FireTV".

From a project manager I would appreciate a less hostile attitude towards users.

**Comment by alekseyzimin** [comment]
No, I will not go and buy a different device.  

I like Fire TV, because it has a lot of content and apps that I use.  Much more than Nvidia Shield.  For example National Geographic/ Science Channel apps are non existent on Android TV but they are well implemented on Fire TV. Fire TV is just not as great natively for my local content, like backed-up DVD and blu-ray ISO's etc., as it is not designed for that.  But  Kodi accomplishes the task of playing local content very well.  I can buy a separate device for Kodi use, but this would be another box with wires connected to my TV, and this is completely unnecessary, as Fire TV hardware is completely capable of playing all my content well enough for my living room TV.  For advanced Kodi use I have libreelec S905X box in my basement home theater system.
  
Why one would want to restrict capabilities of Kodi is beyond my understanding. This looks like team-Kodi is becoming politcal in its decisions instead of user-oriented.  I am all for clean code, reliability and compatibility, but hiding useful and well implemented options like SW decoding from users is unwise.

**Comment by fritsch** [comment]
The only way to help Android is not to add workarounds for every shitty 40 dollar device someone spits out for the sole goal to sell users their own content.

Why do you give amazon money and don't expect from them that they fix their product? On the other hand you argue with volunteers that they add workarounds?

2 for AMLogic
3 for Mediatek
2 for FireTV1
3 for FireTV3
1 for FireTV 4K

and so on? Is that really the strategy that you want from an OSS mediacenter like kodi with exactly one Android Developer?

Buy a 40 dollar device from amazon and expecting that volunteers fix their shit in the freetime is unwise ...

**Comment by Syxton** [comment]
I don't think I'm the only person that believes Amazon as a company has more devices out there (read a larger base for Kodi installs) than AMLogic.  So maybe not every workaround is a required one, but who drew the line on Amazon devices?  Seems really foolish.  You can have a very clean code, but if a large portion of the user base moves to a competitor that DOES work for their device, what good is that strategy?  https://bestdroidplayer.com/streaming-apps/best-kodi-alternatives/ it is much cheaper to swith software than to switch hardware.  I myself have 7 fire devices that run Kodi.

**Comment by fritsch** [comment]
Let me ask differently: Which workarounds for which device do you find here:  https://github.com/xbmc/xbmc/blob/master/xbmc/cores/VideoPlayer/DVDCodecs/Video/DVDVideoCodecAndroidMediaCodec.cpp

"It is much cheaper to switch software than switch hardware" <- I believe, that this is cheaper for you - as it's my freetime that is wasted to write the code.

**Comment by fritsch** [comment]
Here, have fun with your 7 devices: http://mirrors.kodi.tv/test-builds/android/arm/kodi-20190220-4666efbf-androidswmpeg2-armeabi-v7a.apk

**Comment by Syxton** [comment]
If Kodi wants to have pure code, they have to offer their own hardware and only support that (Apple style).  However, by its own description it is made to work on many devices, so "workarounds" are part of the deal.  I guess that a totally pure codebase that works on only select devices is what you want.  As a developer myself, I can't help but shake my head at how this thread has gone.  It is a sad day in the life of this project.

**Comment by MartijnKaijser** [comment]
If you are a developer then you should step up and start fixing stuff instead of demanding workarounds for bad devices.

**Comment by fritsch** [comment]
It's a sad day for you as developer, demanding workaround stuff from OSS developers while at the same being able to fix the root cause yourself, so that no workarounds would have been needed, this is really sad, especially as your highly wanted workaround is just 10 lines of code ...
